2018/socrates-linz/README.md

```diff
@@ -0,0 +1,49 @@
+
+# Mutation-testing
+
+- https://en.wikipedia.org/wiki/Mutation_testing -> `Mutation testing (or mutation analysis or program mutation) is used to design new software tests and evaluate the quality of existing software tests. Mutation testing involves modifying a program in small ways. Each mutated version is called a mutant and tests detect and reject mutants by causing the behavior of the original version to differ from the mutant. This is called killing the mutant. Test suites are measured by the percentage of mutants that they kill.`
+- https://github.com/zimmski/go-mutesting my implementation for Go
+- Run `go-mutesting --debug github.com/symflower/sessions/2018/socrates-linz/csvler` -> `The mutation score is 0.333333 (3 passed, 6 failed, 3 duplicated, 0 skipped, total is 9)` -> There are just 3 "mutations" that are tested by our unit-tests but 6 cases are not! (3 are duplicates -> go-mutesting did some work that lead to the same program/coverage/results)
+
+# Fuzzing
+
+- https://en.wikipedia.org/wiki/Fuzzing -> `Fuzzing or fuzz testing is an automated software testing technique that involves providing invalid, unexpected, or random data as inputs to a computer program.`
+- My implementation of a fuzzing/model-based testing and delta-debugging tool in one binary: https://github.com/zimmski/tavor The file format is defined here https://github.com/zimmski/tavor/blob/master/doc/format.md (there is also an API).
+- Show "example.tavor"
+- Run `tavor --format-file example.tavor fuzz`
+- Run `tavor --format-file example.tavor fuzz --strategy AllPermutations`
+- Show "csvler.tavor"
+- Do the csvler.tavor with just two columns -> not likely to run into the bug
+- Run `tavor --format-file csvler.tavor --max-repeat 10 fuzz` -> generates randomly CSV files with our format. And repeats things at a maximum 10 times.
+- TODO There is a bug in Tavor where we cannot use +0,$cc.Value("," [1-9]) because the [1-9] item gets copied and not newly generated
+- Run `bash fuzz.bash`
+
+# Model-based testing
+
+- https://en.wikipedia.org/wiki/Model-based_testing -> `Model-based testing is an application of model-based design for designing and optionally also executing artifacts to perform software testing or system testing. Models can be used to represent the desired behavior of a system under test (SUT), or to represent testing strategies and a test environment.`
+- Model-based testing is not like fuzzing just about data... it also about how you execute the system under test (SUT) with that data and how to validate that there is a problem or not.
+- E.g. https://github.com/zimmski/tavor/blob/master/doc/complete-example.md -> coin(puts something in) and credit (checks that the credit is now correct)
+- E.g. website checker: Home -> Login -> Profile -> Change Name (these are all actions on the website) the validation would then be something like is there a response status 200 OK.
```

Comment:
would then be something like is there a response status 200 OK. -> would then for instance check whether the response status is 200 OK.